MAVEN MIGRATION - ZOOKEEPER-3028 - Create maven assembly

nkalmar opened the pull request.

eolivelli commented on src/assembly/jar.xml:
```diff
@@ -0,0 +1,65 @@
+          xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+          xsi:schemaLocation="http://maven.apache.org/ASSEMBLY/2.0.0 http://maven.apache.org/xsd/assembly-2.0.0.xsd">
+
+  <id>jar</id>
```
Why do we need this file?
We are splitting the project in sub modules, it is better to have one jar per module.

I guess you are trying to reproduce the ant based build but it is not necessary to me.

Maven has already its clear way of distributing the artifacts.

Or maybe I am missing some part of the story

nkalmar replied:
Yes, that was the goal. Originally, we said we will reproduce ant build as much as possible. It has this fatjar in it.

I will do some revision on this PR.

nkalmar commented:
retest this please

anmolnar commented:
@nkalmar @eolivelli 
I'm trying to test this patch by running `mvn release:prepare` and got the following error message:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-release-plugin:2.5.3:prepare (default-cli) on project zookeeper: Cannot prepare the release because you have local modifications :
[ERROR] [zookeeper-server/src/test/resources/data/invalidsnap/version-2/snapshot.83f:deleted]
```

But I don't have any local modifications, it was created by Maven for some reason.

anmolnar commented:
It was my fault, I got it working now.

I'm trying to compare the artifacts generated by this process: https://cwiki.apache.org/confluence/display/ZOOKEEPER/HowToRelease
and by this patch.

(btw HowToRelease page is still based on SVN??)

Looks like the artifacts are slightly different: Maven build included some text files and scripts and also the `lib` folder in the main jar. Similar differences are experienced in other jars too.

eolivelli commented:
@anmolnar fit is better to rebase this patch to current master because maven build was not working due to the lack of commons-lang.

I think the using the maven release needs some additional work and it can be addressed in a separate patch.

We should have consensus about the way we are going to release the new artifact, see the mail on the mailing list.

With this patch I think you will be able to create the final package just be issuing this command
mvn clean package -DskipTest 

and not mvn release:prepare, it will perform additional work and it will try to push stuff on git repository

A common release procedure with the maven release plugin is like this one:
https://github.com/diennea/herddb/wiki/Release-guide

anmolnar commented:
@eolivelli Rebased on current master and tried with the command you suggested.

Now the artifacts contain "SNAPSHOT" in their name and pretty much the same issue with the contents: scripts, additional files are includes and lib folder in the main jar.

eolivelli commented:
This is expected.
During release the plugin will:
- create release branch
- change versions
- commit and push
- build
- push artifacts to a staging repository
- create tag
- advance version in pom for the new SNAPSHOT version

Enrico

eolivelli commented:
@anmolnar I will try locally with this patch and create a stub for  release procedure.

We should decide what is the expected result, see the mail on the mailing list

nkalmar commented:
Yes, currently the Assembly plugin makes the "maven tarball" which is called in the package phase. 
When testing the release plugin, use -DdryRun=true, so it will not try to upload artifacts. 
I used "mvn release:prepare -DdryRun=true" for testing. It will create a bunch of extra files, you can delete them with git clean -fdx , mvn release:rollback should also work. 

For calling the assembly, I just did what Enrico mentioned "mvn clean package -DskipTest"

I will update the jira with these information.

There are no more maven patches planed, only the fixes.

anmolnar commented:
Thanks @nkalmar @eolivelli will try that. Btw you're not on the same page regarding further PR. Norber said this should be the last one, Enrico expects more.

Also, you did not say anything about the contents of jars and why they're different from the one generated by Ant.

eolivelli commented:
My comment was only because I feel we will need some fine tuning as soon as the rest of the community will evaluate the new release model.

Maybe we can be done with this patch, it will be great.

anmolnar commented:
Hi @eolivelli @nkalmar .
I did another test today by rebasing this patch on latest master and running `mvn clean package -DskipTests` command. I looked at the following artifacts:

**Source package tar.gz and zip in target folder**
*We release this as the source package according to Apache standards (https://www.apache.org/legal/release-policy.html#source-packages)*

Issues:
- root level `pom.xml` file is missing and the project cannot be built,
- recipes subprojects are missing, so even if I copy over the top-level pom.xml, the project cannot be built,
- zookeeper-contrib project is missing: we don't want to include in the release?
- mvn install is looking for `src/assembly/source-package.xml`, but it's missing
- also looking for `.git` directory in zookeeper-server which is obviously missing,
- overall, the user MUST be able to build it without any modifications, so this needs to be fixed.

**Maven jars in target folders (sources, tests, main)**

- main jar: `lib` folder and resource files,
- tests.jar: `data` fodler and resource files, patch build / github scripts,
- sources.jar: `lib` folder and resource files.

There're 2 additional jar files created by the Ant process: `bin` and `javadoc` which I haven't found in the maven build.

nkalmar commented:
Fixed the tarball. I tested by unzipping to a completely clear directory, the build should be working now.

I will take a look on the jars tomorrow.

anmolnar commented:
Thanks @nkalmar !
The build works as expected now. However that top level `src` directory is a little bit ugly. Can we put that single file somewhere else?

nkalmar commented:
Yes, it is ugly... that's where it should be according to maven standard library structure, but we can put it wherever we want. 
How does conf/assembly sounds instead of src/assembly?

nkalmar commented:
I moved the file, for now, it's as good as anywhere else.

tamaashu commented:
I'm not sure if I would put it in **conf**. That directory is for runtime configuration.
It might be better in the root directory. Not perfect, but still better...

nkalmar commented:
Allright, moved it.

nkalmar commented:
Fixed the jars, added javadoc jar (turned off the errors as it would fail miserably due to lots of missing javadoc - still getting tons of warnings though) 

The bin jar should be the standard jar, that contains the classes.

anmolnar commented:
@nkalmar 
- Jars look good to me now,
- source descriptor xml has been moved to project dir which is fine,
Issues:
- Strange empty jar `zookeeper-3.6.0-SNAPSHOT-tests.jar` has been generated in root /target folder,
- Documentation about the new release process `HowToRelease.md` is still missing and I'd like to have it in this patch

eolivelli commented:
This 'release guide' is maven based, maybe it can help.
I think there is some part missing:
- gpg keys setup
- close the staging repository
- use 'apache-release' profile (maybe not needed)

tamaashu commented:
@nkalmar in the main pom.xml there is a section for committers, could you please add fangmin?
He became committer after that file has been created.

nkalmar commented:
Thanks guys, fixing them now. Also adding @lvfangmin to developers, using his public git profile to fill out the informations needed.

anmolnar commented:
@nkalmar README.md file is not the best place to document the release process with Maven. I see that it already has a section which describes what's _inside_ the release artifacts for the user, but it's nothing to do about creating a new release. In other words, the target audience is different: Readme is for users, while HowToRelease doc is for Release Managers. Please move this doc to the wiki instead, next to https://cwiki.apache.org/confluence/display/ZOOKEEPER/HowToRelease. Make a copy and replace the relevant sections with Maven alternatives.

Also please give it some nice styling to make it more readable.

Still, you guys (@nkalmar , @eolivelli) don't have an agreement on the commands which should be used for testing. `mvn release:prepare` something... or `mvn clean package` blah-blah is the right one? I'm confused.

@eolivelli Which Maven guide are you referring to? You didn't link anything. :(

nkalmar commented:
I will update the confluence page.

When we finalize the release procedure, I will also update the README.md to display the exact end artifact informations.

mvn release prepares and uploads the maven artifacts (jars in this case for the modules) to central repo. package is needed to produce the tarball.

tamaashu commented:
Hi @nkalmar great work.
I like the fully Apache compliant `src` tarball, but for convenience reasons could we also have a `bin` tarball, just like for HBase? https://hbase.apache.org/downloads.html
Sometimes people would not like to build their own ZooKeeper, just download, extract and run it.
Thanks

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/29/
--none----none--

anmolnar commented:
@nkalmar @eolivelli @tamaashu Where are we with this patch? Is it ready to be committed or something is missing?
Source tarballs looks OK to me, but there's also an empty `zookeeper-3.6.0-SNAPSHOT-tests.jar`. I think we don't need that.

Regarding the convenience tarball, I don't see there's is a consensus in the community, so we can leave it for a later patch if needed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/<h2>Failed Tests: <span class='status-failure'>12</span></h2><h3><a name='PreCommit-ZOOKEEPER-github-pr-build-maven/org.apache.zookeeper:zookeeper-server' /><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport'>PreCommit-ZOOKEEPER-github-pr-build-maven/org.apache.zookeeper:zookeeper-server</a>: <span class='status-failure'>12</span></h3><ul><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testInconsistentPeerType/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testInconsistentPeerType</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testLeaderOutOfView/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testLeaderOutOfView</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testMetricsProviderConfiguration/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testMetricsProviderConfiguration</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testInvalidMetricsProvider/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testInvalidMetricsProvider</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testQuorumDefaults/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testQuorumDefaults</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testFaultyMetricsProviderOnStop/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testFaultyMetricsProviderOnStop</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testMetricsProviderLifecycle/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testMetricsProviderLifecycle</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testFaultyMetricsProviderOnConfigure/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testFaultyMetricsProviderOnConfigure</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testElectionFraud/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testElectionFraud</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testBadPeerAddressInQuorum/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testBadPeerAddressInQuorum</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testFaultyMetricsProviderOnStart/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testFaultyMetricsProviderOnStart</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/30/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.test/ReadOnlyModeTest/testSeekForRwServer/'><strong>org.apache.zookeeper.test.ReadOnlyModeTest.testSeekForRwServer</strong></a></li></ul>
--none----none--

nkalmar commented:
Hmm, this log4j error come with the rebase. Will look into it.

nkalmar commented:
@anmolnar 
So there weren't 100% agreement on how the release should look like, and unfortunately not many response came on dev list either. Basically us active on this PR. So, from my side, it sounds good to go with buildable source tarball + apache profile release. We can add the "convenience tarball" any time. 

I will look into that empty jar...

anmolnar commented:
@nkalmar So this patch can be merged as it is if I understood you correctly. (once you removed the empty jar)

nkalmar commented:
Yes, and I just removed the empty jar generation

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/<h2>Failed Tests: <span class='status-failure'>12</span></h2><h3><a name='PreCommit-ZOOKEEPER-github-pr-build-maven/org.apache.zookeeper:zookeeper-server' /><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport'>PreCommit-ZOOKEEPER-github-pr-build-maven/org.apache.zookeeper:zookeeper-server</a>: <span class='status-failure'>12</span></h3><ul><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testInconsistentPeerType/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testInconsistentPeerType</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testLeaderOutOfView/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testLeaderOutOfView</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testMetricsProviderConfiguration/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testMetricsProviderConfiguration</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testInvalidMetricsProvider/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testInvalidMetricsProvider</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testQuorumDefaults/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testQuorumDefaults</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testFaultyMetricsProviderOnStop/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testFaultyMetricsProviderOnStop</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testMetricsProviderLifecycle/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testMetricsProviderLifecycle</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testFaultyMetricsProviderOnConfigure/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testFaultyMetricsProviderOnConfigure</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testElectionFraud/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testElectionFraud</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testBadPeerAddressInQuorum/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testBadPeerAddressInQuorum</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.server.quorum/QuorumPeerMainTest/testFaultyMetricsProviderOnStart/'><strong>org.apache.zookeeper.server.quorum.QuorumPeerMainTest.testFaultyMetricsProviderOnStart</strong></a></li><li><a href='https://builds.apache.org/job/PreCommit-ZOOKEEPER-github-pr-build-maven/32/org.apache.zookeeper$zookeeper-server/testReport/org.apache.zookeeper.test/ReadOnlyModeTest/testSeekForRwServer/'><strong>org.apache.zookeeper.test.ReadOnlyModeTest.testSeekForRwServer</strong></a></li></ul>
--none----none--

nkalmar commented:
So this maven test fail still remains after backport, I might have merged a pom.xml incorrectly... need to check. Sadly rebase is followed by force push :(

nkalmar commented:
Why is the build failing on javadocs warnings? Those are old warnings. I can fix them, it's just strange.

nkalmar commented:
retest this please

tamaashu commented:
retest this please

eolivelli commented:
@nkalmar  @anmolnar  @tamaashu 

in order to restart Travis you have to open travis and click on the 'Restart build' button.
I guess only the owner of the PR can do it.

nkalmar commented:
retest this please

anmolnar commented:
Committed to master. Thanks @nkalmar !
